### PR TITLE
Update the standalone flag in automm.predictor.save() to to save the pertained  model for offline deployment

### PR DESCRIPTION
**Pull request by yiqings**
*Issue #, if available: https://github.com/awslabs/autogluon/issues/1115#issuecomment-1048434359

*Description of changes: An efficient implementation to save the donwloaded models from transformers for the offline deployment. We @sxjscience revise the logics in our previous PR https://github.com/awslabs/autogluon/pull/1572, as discussed in https://github.com/awslabs/autogluon/pull/1572#issuecomment-1051014878.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: text/src/autogluon/text/automm/predictor.py
Comment (sxjscience) [inline text/src/autogluon/text/automm/predictor.py:875]: Consider to change `'LOCAL:'` to `'local://'`.

File: text/src/autogluon/text/automm/predictor.py
Comment (sxjscience) [inline text/src/autogluon/text/automm/predictor.py:945]: Change to `model_config.checkpoint_name = os.path.join(path, model_config.checkpoint_name[len('local://'):])`

File: text/src/autogluon/text/automm/predictor.py
Comment (sxjscience) [inline text/src/autogluon/text/automm/predictor.py:873]: One approach is to implement the functionality as follows:
```python
if model_name == 'clip' or model_name == 'hf_text':
   model_config = getattr(self._config.model, model_name)
   checkpoint_name = model_config.checkpoint_name

   if checkpoint_name.startswith('local://'):
      new_checkpoint_name = checkpoint_name
      standalone_save_path = checkpoint_name[len('local://'):]
   else:
      new_checkpoint_name = os.path.join('local://', model_name)
      standalone_save_path = os.path.join(path, model_name)

   AutoConfig.from_pretrained(checkpoint_name).save_pretrained(standalone_save_path)
   AutoTokenizer.from_pretrained(checkpoint_name).save_pretrained(standalone_save_path)
   AutoModel.from_pretrained(checkpoint_name).save_pretrained(standalone_save_path)
   model_config.check_point_name = new_checkpoint_name
```

**Comment by sxjscience** [comment]
Need to add a test-case about `predictor.save(standalone=True)` in https://github.com/awslabs/autogluon/blob/master/text/tests/unittests/text_prediction/test_predictor_pytorch.py

**Comment by szha** [comment]
Job PR-1575-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/2/index.html

**Comment by szha** [comment]
Job PR-1575-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/3/index.html

**Comment by szha** [comment]
Job PR-1575-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/4/index.html

**Comment by szha** [comment]
Job PR-1575-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/5/index.html

**Review by sxjscience — APPROVED** [review]

File: text/tests/unittests/automm/test_automm_predictor.py
Comment (sxjscience) [inline text/tests/unittests/automm/test_automm_predictor.py:242]: @zhiqiangdon We disabled network access via the mock.

Comment (sxjscience) [reply text/tests/unittests/automm/test_automm_predictor.py:242]: The solution is borrowed from https://stackoverflow.com/questions/18601828/python-block-network-connections-for-testing-purposes

**Review by sxjscience — DISMISSED** [review]
After offline discussion with @zhiqiangdon , we should also download timm to local.

**Comment by sxjscience** [comment]
@yiqings Yeah, we may add the logic of saving timm to local. After that, it's fine to merge the PR.

**Comment by sxjscience** [comment]
@yiqings You may consider to add an option to change this flag: https://github.com/awslabs/autogluon/blob/297117d448f937f384fd5a79326fd99bdc62d068/text/src/autogluon/text/automm/models/timm_image.py#L49

**Comment by Innixma** [comment]
@sxjscience Status of this PR? v0.4.0 is planned for March 9th. Can we test and merge this PR by March 8th? If not we should wait until after release to merge this.

**Comment by sxjscience** [comment]
This feature won't be targeted for 0.4.

**Comment by szha** [comment]
Job PR-1575-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/8/index.html

**Review by sxjscience — APPROVED** [review]
LGTM. The test-case also looks good.

**Comment by sxjscience** [comment]
@zhiqiangdon Would you also review it?

@yiqings Since we are in code freeze, this feature won't be merged in the 0.4 release and we will merge it after we have released 0.4.

File: text/src/autogluon/text/automm/predictor.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:1004]: Is this block redundant? All the weights are loaded later in this function.

File: text/src/autogluon/text/automm/predictor.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:867]: Two typos of standalong.

File: text/src/autogluon/text/automm/models/timm_image.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/models/timm_image.py:52]: If the checkpoint name has a local path, we may need to load the checkpoint after the model is created.

**Comment by zhiqiangdon** [comment]
> @zhiqiangdon Would you also review it?
> 
> @yiqings Since we are in code freeze, this feature won't be merged in the 0.4 release and we will merge it after we have released 0.4.

Added some comments. I think we may need to adjust the logic of loading a local timm checkpoint.

File: text/src/autogluon/text/automm/predictor.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:969]: Ditto.

File: text/src/autogluon/text/automm/predictor.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:888]: Consider wrapping this into a function and put it in utils.py. In this way, the logic is clear in this function.

File: text/src/autogluon/text/automm/predictor.py
Comment (yiqings) [reply text/src/autogluon/text/automm/predictor.py]: Thanks. Already wrapped by `load_pretrained_configs' in utils.py.

File: text/src/autogluon/text/automm/predictor.py
Comment (yiqings) [reply text/src/autogluon/text/automm/predictor.py]: Thanks. Already wrapped by `save_pretrained_model' in utils.py.

File: text/src/autogluon/text/automm/predictor.py
Comment (yiqings) [reply text/src/autogluon/text/automm/predictor.py]: Thanks, the typos are now corrected.

File: text/src/autogluon/text/automm/utils.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/utils.py:539]: It's unclear why do we save a pretrained timm checkpoint in a 'pkl' file.

File: text/src/autogluon/text/automm/utils.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/utils.py:566]: Where do we use 'pytorch_model.bin' and 'config.json'?

Comment (sxjscience) [reply text/src/autogluon/text/automm/utils.py:566]: I think we are safe to remove the check here.

File: text/src/autogluon/text/automm/utils.py
Comment (sxjscience) [reply text/src/autogluon/text/automm/utils.py:539]: I think we are safe to remove the check here.

Comment (sxjscience) [reply text/src/autogluon/text/automm/utils.py:539]: Can we save to `.pt` ?

File: text/src/autogluon/text/automm/predictor.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:976]: I think we can just set pretrained=False, since we are loading from a trained checkpoint saved by AutoMM.

File: text/src/autogluon/text/automm/utils.py
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/utils.py:539]: timm uses `.pth`.

File: text/src/autogluon/text/automm/utils.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/utils.py:545]: Overall, this function mainly converts local path to global path. Adding `pretrained` in `config` also looks kind of confusing without deep understanding. I think we can avoid the path conversion by adding a new `local_path`  key under the hf_text and clip configs. Basically, we can do `checkpoint_name=os.path.join(local_path, checkpoint_name)` before passing `checkpoint_name` to HFAutoModelForTextPrediction and CLIPForImageText.

Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/utils.py:545]: If `local_path ` is None, it means downloading from Huggingface. So, no need additional flag `pretrained` to indicate that. Adding `local://` is also unnecessary f using `local_path`.

File: text/src/autogluon/text/automm/utils.py
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/utils.py:533]: It seems that saving a pretrained timm checkpoint in `.save()` is not necessary. We can just `pretrained=False` when calling `create_model` in `.load()`. No model downloading is required for timm.

Comment (sxjscience) [reply text/src/autogluon/text/automm/utils.py:533]: +1. @yiqings need to fix it.

File: text/src/autogluon/text/automm/predictor.py
Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py:976]: +1. @yiqings need to fix it.

File: text/src/autogluon/text/automm/utils.py
Comment (yiqings) [reply text/src/autogluon/text/automm/utils.py:539]: Great! Thanks for the advice. It is suggested no formatting difference between `.pkl`, `.pt` and `.pth`, except for a difference in suffix, eg [https://blog.karatos.in/a?ID=01600-da733521-cc0a-4be1-a696-41930451a6a9](url). We are going to use `.pth` to get aligned with timm.

File: text/src/autogluon/text/automm/utils.py
Comment (yiqings) [reply text/src/autogluon/text/automm/utils.py:566]: Both 'pytorch_model.bin' and 'config.json' are used by calling `create_model` in AutoMMPredictor.load with a checkpoint referring to the local dir where they are stored. To be more specific, in `create_model` of AutoMMPredictor.load  [https://github.com/yiqings/autogluon/blob/8985d83bf238432ebba5062655a7a4b45bea44d3/text/src/autogluon/text/automm/predictor.py#L971](url), config.model.{MODEL NAME}.checkpoint_name is a local dir pointing to 'pytorch_model.bin' and 'config.json'.

Comment (yiqings) [reply text/src/autogluon/text/automm/utils.py:566]: The assertion guarantees the existence of 'pytorch_model.bin' and 'config.json'. The missing of one of them will raise an error in loading the models in `create_model`.

File: text/src/autogluon/text/automm/predictor.py
Comment (yiqings) [reply text/src/autogluon/text/automm/predictor.py:976]: We follow the original timm settings by setting pretained =True for downloading online pretrained weights, c.f. [https://github.com/rwightman/pytorch-image-models/blob/7c67d6aca992f039eece0af5f7c29a43d48c00e4/timm/models/factory.py#L37](url). Here is the annotations for our `create_model` [https://github.com/yiqings/autogluon/blob/8985d83bf238432ebba5062655a7a4b45bea44d3/text/src/autogluon/text/automm/utils.py#L427](url), in which we call TimmAutoModelForImagePrediction in ./text/src/autogluon/text/automm/models/timm_image.py, and its annotation reads [https://github.com/yiqings/autogluon/blob/8985d83bf238432ebba5062655a7a4b45bea44d3/text/src/autogluon/text/automm/models/timm_image.py#L47](url).

Comment (yiqings) [reply text/src/autogluon/text/automm/predictor.py:976]: Conclusively, by keeping aligned with timm, we write pretained =True for online downloading pretrained weights, and pretained =False for using the local pretrained weights. Probably it does cause same confusion, we may change the name of this flag.

File: text/src/autogluon/text/automm/utils.py
Comment (yiqings) [reply text/src/autogluon/text/automm/utils.py:533]: Is it because timm models weights are already saved in model.ckpt file?

File: text/src/autogluon/text/automm/utils.py
Comment (yiqings) [reply text/src/autogluon/text/automm/utils.py:545]: Thanks for the great comments. I'll try into incorporate them.

File: text/src/autogluon/text/automm/predictor.py
Comment (yiqings) [reply text/src/autogluon/text/automm/predictor.py:976]: Is it because timm models weights are already saved in model.ckpt file?

Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/predictor.py:1025]: Yes, that's the weights we load in `.load()`. Timm supports not downloading checkpoints if `pretrained=False`.

**Comment by szha** [comment]
Job PR-1575-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/13/index.html

File: text/src/autogluon/text/automm/utils.py
Comment (yiqings) [reply text/src/autogluon/text/automm/utils.py]: This block is now removed in [https://github.com/awslabs/autogluon/pull/1575/commits/b02bebb383d4e004d00380814baef9293977aea4](url)

**Comment by szha** [comment]
Job PR-1575-17 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/17/index.html

**Comment by szha** [comment]
Job PR-1575-18 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1575/18/index.html

**Review by zhiqiangdon — APPROVED** [review]

**Pull request merged**
